### PR TITLE
fix: feature to provide flexibility in time selection in availability page.

### LINKED ISSUE #15763: Availability doesnt allow 5 minute steps only 15 minute steps
### Issue Summary
Currently the Availability / Working Hours only allow the usage of 15 Minute steps, so for example:
image

We have some usecases where we would need though 16:10 or 16:05. We would prefer if:
- 5 Minute Steps are choosable
- manual typing orf "16:05" gets accepted

### Steps to Reproduce
1. go to availability
2. add working hours of "08:20 - 11:50" and "13:00 - "16:05"

### Expected Results
- More flexibility on the working hours to respect opening times of local shops / childcare that are not always in 15min steps.

### PR DESCRIPTION
**Pull request by vijayraghav-io**
## What does this PR do?

- Fixes #15763 
- Fixes CAL-

**Please NOTE: This does not affect selection of time slots in booking page. This affects only time selection in availability setup Page.**

https://www.loom.com/share/1611df99556b47f9983002868a91317b?sid=4ad0573e-8c72-4db4-a73c-f081c88e93b8

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by masteradhoc** [comment]
Thanks for the work here Vijay and the video to take a closer look! Looks good and closes the issue noted on #15763.
As soon as this is published this will allow businesses to be much more flexible in selection of opening times.
Still the solution you did doesnt compromise the clean and easy UI due to the dropdown selection on the right side.

Nice Solution and work here! Thanks.

File: apps/web/public/static/locales/en/common.json [hunk 1]
```diff
@@ -2569,9 +2569,11 @@
...
   "booking_question_response_variables": "Booking question response variables",
   "managed_by_teamAdmins": "Managed by {{teamAdmins}}",
-  "number_of_options":"{{count}} options",
+  "number_of_options": "{{count}} options",
   "reschedule_with_same_round_robin_host_title": "Reschedule with same Round-Robin host",
   "reschedule_with_same_round_robin_host_description": "Rescheduled events will be assigned to the same host as initially scheduled",
-   "disable_input_if_prefilled": "Disable input if the URL identifier is prefilled",
+  "disable_input_if_prefilled": "Disable input if the URL identifier is prefilled",
+  "time_increment": "Time increment",
+  "time_increment_hint": "Time increment affects the time selection options in this page only, will not affect the way time displayed in booking page.",
```
Comment (Amit91848) [inline apps/web/public/static/locales/en/common.json:2577]: What does this even mean?

**Review by Amit91848 — CHANGES_REQUESTED** [review]
https://github.com/user-attachments/assets/12360c02-046c-475f-afbf-8b29d2b7f5a7

So the general starting for each day is 9:00. Where first slot is 9:00 and second is 9:15

I changed the starting for Monday from 9:00 to 9:05. Now instead of starting from 9:05, it just skipped the first slot of 9:00 - 9:15 as it falls outside of availability, which is correct. As per my understanding of the issue, the slots should be 9:05, 9:20, 9:35, etc?? Feel free to chime in @masteradhoc as I am not completely sure what you expect the slots to look like on booking page.

But here it just skips the first slot of 9:00 - 9:15 and shows the second slot starting from 9:15. Which can already be achieved by using the starting of 9:15 with the current 15 minutes steps.

**Comment by vijayraghav-io** [comment]
> Recording.2024-08-25.162444.mp4 
> So the general starting for each day is 9:00. Where first slot is 9:00 and second is 9:15
> 
> I changed the starting for Monday from 9:00 to 9:05. Now instead of starting from 9:05, it just skipped the first slot of 9:00 - 9:15 as it falls outside of availability, which is correct. As per my understanding of the issue, the slots should be 9:05, 9:20, 9:35, etc?? Feel free to chime in @masteradhoc as I am not completely sure what you expect the slots to look like on booking page.
> 
> But here it just skips the first slot of 9:00 - 9:15 and shows the second slot starting from 9:15. Which can already be achieved by using the starting of 9:15 with the current 15 minutes steps.

The 15slot shown in booking page can be changed in the eventType settings.

**Comment by Amit91848** [comment]
> The 15slot shown in booking page can be changed in the eventType settings.
> 
> https://www.loom.com/share/7584183c8c0b403680f43f111aba6761?sid=af466689-0555-47f3-b8a3-b09cd0ad2a1a

I am not talking about the duration of the event. If your event type duration is 15 minutes and starting time is 9:05, shouldn't the slots start from 9:05 and be 9:05 - 9:20, 9:20 - 9:35, 9:35-9:50, etc.

But when you start the slots from 9:15 onwards, wouldn't it just make sense to select the starting time as 9:15 only. Which begs the question why do we need this pr in the first place?

**Comment by vijayraghav-io** [comment]
> I am not talking about the duration of the event. If your event type duration is 15 minutes and starting time is 9:05, shouldn't the slots start from 9:05 and be 9:05 - 9:20, 9:20 - 9:35, 9:35-9:50, etc.
> 
> But if you start the slots from 9:15 onwards, wouldn't it just make sense to select the starting time as 9:15 only. Which begs the question why do we need this pr in the first place?

The scenario you are saying is correct for 15mins eventtype, it should start from 9:15.
But if the event type is 10mins or 20min, the user can't define their availability exactly because they can select only 15min increments in availability page. So by allowing the user to select say 10mins in availability page, they can define their availability accurately and show their available slots accurately for such events.

**Comment by masteradhoc** [comment]
What i saw in the first Video is correct as you can insert your opening times more exact lets say by 09:10 to 09:40. So far so good.

For booking my events (f.e. 15 Minute Intro Calls) it should be possible to make these bookings:
- 09:10 - 09:25
- 09:25 - 09:40

So the 15'/30' Booking times should respect the openinghours.

**Comment by vijayraghav-io** [comment]
Yes it's not the scope of this PR to change the booking slots logic.

The flexibility of selecting time availability provided by this PR coupled with event type duration selections will allow for more accurate display of booking time slots for specific cases.

**Comment by Amit91848** [comment]
> The scenario you are saying is correct for 15mins eventtype, it should start from 9:15. But if the event type is 10mins or 20min, the user can't define their availability exactly because they can select only 15min increments in availability page. So by allowing the user to select say 10mins in availability page, they can define their availability accurately and show their available slots accurately for such events (of 5mins or 10mins or 20mins. 35.. 40.. 50...).

Yes but it would still be working only for some combinations of durations and time increments and breaking / not accurate for others. Which is why I was concerned.

> Seems like we should split the 15min only being able to start at :15 / :30 / :45 AND :00 to a seperate task. Solutions are already in the work based on the comments on #16307.
> 
> Therefore the PR seems to fix the issue i created. Looks good to me.

But now it makes sense, I missed that conversation. Will do another review tomorrow @vijayraghav-io @masteradhoc

**Comment by masteradhoc** [comment]
@Amit91848 definately agree on you for that one. i think this could be merged and the other PR / Solution that was noted could be worthed checked afterwards. Would make cal.com much more flexible!

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
We might just want to allow 5-minute time increments by default, I'll see what the team thinks about that and come back to you 🙏

**Comment by CarinaWolli** [comment]
> Seems like we should split the 15min only being able to start at :15 / :30 / :45 AND :00 to a seperate task.
> Solutions are already in the work based on the comments on #16307.

We are still discussing this internally, but we will most likely add a new setting that allows you to define different slot start times than our default ones.

**Comment by vijayraghav-io** [comment]
> We might just want to allow 5-minute time increments by default, I'll see what the team thinks about that and come back to you 🙏

Thankyou!  @CarinaWolli for looking into this or reviewing this, will wait for your follow-up / decison.
Just to put my thoughts - By changing the value of env variable to 5mins from 15mins,  may affect all the users, and thought may not be user friendly for all type of users if they want to change. 
Also this gives flexibility to keep increment tied to a availability. Availability A can have 5min increment, B can have 15min like so..

Anyhow will leave the final call to team and implement accordinly.

**Comment by CarinaWolli** [comment]
> Thankyou! @CarinaWolli for looking into this or reviewing this, will wait for your follow-up / decison. Just to put my thoughts - By changing the value of env variable to 5mins from 15mins, may affect all the users, and thought may not be user friendly for all type of users if they want to change. Also this gives flexibility to keep increment tied to a availability. Availability A can have 5min increment, B can have 15min like so.. And having this as user setting will enforce same increment for all availabilities.
> 
> Anyhow will leave the final call to team and implement accordingly.

Thanks for your input 🙏 We probably want something that allows the user to type any start or end time for their working hours but the select still shows only options in 15 minute intervals to be more user friendly.

**Comment by masteradhoc** [comment]
@CarinaWolli any update on the status?

**Comment by CarinaWolli** [comment]
> We probably want something that allows the user to type any start or end time for their working hours but the select still shows only options in 15 minute intervals to be more user friendly.

That's exactly what we want, I also updated the issue. I will close this PR, thank you for your effort 🙏

**Comment by vijayraghav-io** [comment]
> > We probably want something that allows the user to type any start or end time for their working hours but the select still shows only options in 15 minute intervals to be more user friendly.
> 
> That's exactly what we want, I also updated the issue. I will close this PR, thank you for your effort 🙏

@CarinaWolli , can you please look into this PR https://github.com/calcom/cal.com/pull/16456, 
I have implemented rounded corners as per your comment. Also  I have addressed and implemented all review comments. I guess there is no any blocking comments. Can you please merge 🙏 ,
Since lot of refactoring are happening in codebase, it will be hard and cumbersome to keep up to date 🙏

**Pull request closed without merging**